cmd/maya-apiserver/spc-watcher/lease.go

```diff
@@ -0,0 +1,109 @@
+/*
+Copyright 2018 The OpenEBS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package spc
+
+import (
+	"fmt"
+	apis "github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+	clientset "github.com/openebs/maya/pkg/client/generated/clientset/internalclientset"
+	env "github.com/openebs/maya/pkg/env/v1alpha1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"strings"
+)
+
+const (
+	// SpcLeaseKey is the key that will be used to acquire lease on spc object.
+	// It will be present in spc annotations
+	// If key has an empty value, that means no one has acquired a lease on spc object.
+	SpcLeaseKey = "openebs.io/spc-create-lease"
+	// PodNameEnvKey is the key to fetch name of the pod,which will be combined with PodNameSpaceEnvKey
+	// to be used as a value of SpcLeaseKey.
+	// e.g. "openebs.io/spc-lease":"openebs/maya-apiserver-6b4695c9f8-nbwl9"
+	PodNameEnvKey = "OPENEBS_MAYA_POD_NAME"
+	// PodNameSpaceEnvKey is the key to fetch the namespace of the pod
+	PodNameSpaceEnvKey = "OPENEBS_NAMESPACE"
+)
+
+// Leases is an interface which assists in getting and releasing lease over an spc object
+type Leases interface {
+	// GetLease will try to get a lease on spc, in case of failure it will return error
+	GetLease() (string, error)
+	// UpdateLease will update the lease value of the spc
+	UpdateLease(leaseValue string) (*apis.StoragePoolClaim, error)
+	// RemoveLease will remove the acquired lease on the spc
+	RemoveLease() *apis.StoragePoolClaim
+}
+
+// spcLease is the struct which will implement the Leases interface
+type spcLease struct {
+	// spcObject is the storagepoolclaim object over which lease is to be taken
+	spcObject *apis.StoragePoolClaim
+	// leaseKey is lease key on current storagepoolclaim object
+	leaseKey string
+	// oecs is the openebs clientset
+	oecs clientset.Interface
+}
+
+func (sl *spcLease) GetLease() (string, error) {
+	// Get the lease value.
+	leaseValue := sl.spcObject.Annotations[sl.leaseKey]
+	// If leaseValue is empty acquire lease.
+	if strings.TrimSpace(leaseValue) == "" {
+		spcObject, err := sl.UpdateLease(sl.getPodName())
+		if err != nil {
+			return "", err
+		}
+		return spcObject.Annotations[sl.leaseKey], nil
+	}
+	// If leaseValue is not empty, lease cannot be acquired.
```

Comment:
What happens if the lease is taken by m-apiserver and it crashes before completing the SPC creation process? Sounds like we will need a lease-expiry mechanism, probably based on the time/presence of the pod that took the lease.

Reply:
When does a resync event get triggered? How often does it get triggered?

Reply:
A resync event will be triggered after a fixed interval of time( for us it is 20 or 30 seconds, not remembering) and we can customize this.
So resync event is basically a update hook that watcher keeps on firing by delevering the old and new spc objects and you can compare the objects etc to  hook in your handlers to deal with the event.

Reply:
cool. So with the current code, what happens in the below scenario:
- SPC is created by kubectl apply -- for  creating pools with disks. 
- m-apiserver gets the lease, but crashes immediately after due to either node failure or another issue. 
- m-apiserver gets re-launched to another node with a new pod name. 
- SPC create event is received by the (new) m-apiserver
--
Need clarification on the following:
(a) It appears that (new) m-apiserver will fail to get lease because the older one was never removed.  To recover from this state, the SPC need to be deleted and re-added. 
(b) When the (new) m-apiserver gets the resync, does it go ahead and create it?

Reply:
**(a) It appears that (new) m-apiserver will fail to get lease because the older one was never removed.  To recover from ---this state, the SPC need to be deleted and re-added.** 

--->Deletion of spc should not be done from maya side IMO. It is a user created resource.
Just how happens PVC -> PV ( Does PV creation failure demands a PVC recreate ? ) 
Our case has bit lengthy chain , SPC->CSP->DEPLOY->SP

**(b) When the (new) m-apiserver gets the resync, does it go ahead and create it?**

--->When the new m-apiserver gets the resync, it does not check for the lease annotation but directly go to resync logic where it figures out current state of world which in our case for now is the pool count. The current pool count for a spc is the count of cstorpool cr that is currently existing.
Then if the desired state of world which is maxPool( specified in spc) is not matching to current state, it will go and call the CreatePool logic ( which has the nodeDiskAlloter mechanism for figuring out the resources to provision the spare amount of pool) and then will try to converge to desired state by creating pools.
So in short resync will always go and try to converge irrespective of spc lease.

But this works out when we have a single instance of maya-pod, for multiple instance there can be point of failure in resync in following case. Let us consider an example wehre maya pod is replicated to a count 2.
1. DesiredPoolCount=6
2. CurrentPoolCount=3
resyncRecieved ---- > mayaPod will try to create 3 spare pool , now if 2 maya pods reads the csp count at same time, all will go and try to converge and we will end having 6 pool resources. This is undesirable !!

But still trying to think when can a resync be faulty even in 1 instance of may pod!

Reply:
Short Story : 
1. Manual Provisioned SPC are not reconciled.
2. SPC lease is significant only in case of create event.
3. Resync will always happen for auto provisioned SPCs by dishonouring the SPC lease.( Problem if maya pod instance>1)
Certainly needs improvement.

Reply:
thank you!  :-)